c++/src/kj/async.c++

```diff
@@ -370,7 +379,7 @@ Promise<void> TaskSet::onEmpty() {
     }
   }
 
-  if (tasks == nullptr) {
+  if (shuttingDown) {
```

Comment:
I noticed this oversight via manual inspection. Added a test to protect against regressing that in the future.

Reply:
What oversight?

This change looks wrong. If you call `onEmpty()` and the task set is already empty, then the promise should complete immediately. That's what the previous code implemented. The new code only returns READY_NOW if the TaskSet is currently being destroyed. If it's empty, but not being destroyed, the new code will incorrectly cause the caller to wait.

Reply:
I think the way it was previously was fine.

1. I doubt anyone calls `onEmpty()` on a `TaskSet` from within its own destructor.
2. If they _were_ doing so before, the behavior already wasn't guaranteed -- `onEmpty()` would only return READY_NOW if nothing else happened to add a task to the TaskSet just prior, during other promise destructors.
3. It we wanted to get pedantic, if `onEmpty()` were to be called as a result of cancelling promises in `~TaskSet()`, it really ought to _never_ return READY_NOW, because the TaskSet isn't actually empty until all the promises have been canceled. So if anything, the test should be `tasks == nullptr && !shuttingDown`. However, I don't think it actually matters.

Reply:
Hmm.... I was just going from the logic from before

```
  while (tasks != nullptr) {
    auto toDestroy = kj::mv(tasks);
    tasks = nullptr;
  }
```

In this case, all the tasks that get destroyed see `onEmpty` return `READY_NOW`. While I agree with you on point #2, I'm worried about Hyrum's law in this case even for Cloudflare's use cases. Specifically, if two tasks enqueued were accidentally using `onEmpty` in their destruction, they would have been fine previously but now they'd be throwing an exception because the first would create a promise fulfiller pair & the second would throw an exception.

Beyond just the exception, it also changes the logic of cancellation. Before, during destruction, the `onEmpty` would get satisfied & any logic triggered as a result would run. If we use `tasks == nullptr && !shuttingDown` then we'll create the `emptyFulfiller` but it'll never get triggered.

Admittedly it's probably a corner case, but I'd rather separate bug fixes & behavioral changes when possible.

Reply:
Again, the old behavior was actually incorrect, and anyone relying on it would presumably be fundamentally broken. Changing them to throw an exception would probably be a good thing, to uncover these bugs.

But also, no one is relying on it. ​It doesn't make sense to call `onEmpty()` on a task set that you don't own, and if you do own the task set, then you wouldn't be calling this after the task set is already being destroyed.

To make extra sure, I went ahead and looked at all call sites of this method. As expected, no one is calling it in a place where the TaskSet could possibly be in the process of being destroyed.

So this commit is trying to maintain buggy behavior that no one actually relies on. Let's just remove it.